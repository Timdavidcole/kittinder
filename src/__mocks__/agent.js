const Breeds = {
  all: jest.fn()
};

const Images = {
  get: jest.fn(),
  getOne: jest.fn()
};

const Favourites = {
  get: jest.fn(),
  post: jest.fn()
};

export default { Breeds, Images, Favourites };
