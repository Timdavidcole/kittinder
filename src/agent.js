import _superagent from "superagent";

const API_ROOT = "https://api.thecatapi.com/v1/";
const API_KEY = "a701ff7c-5713-41b4-b4e7-86671f2ce5a3";

const responseBody = res => res.body;

const requests = {
  get: (url, query) =>
    _superagent
      .get(`${API_ROOT}${url}`)
      .set("x-api-key", API_KEY)
      .query(query)
      .then(responseBody),
  post: (url, body) =>
    _superagent
      .post(`${API_ROOT}${url}`, body)
      .then(responseBody),
  put: (url, body) =>
    _superagent
      .put(`${API_ROOT}${url}`, body)
      .then(responseBody)
};

const Images = {
  all: (query) => requests.get(`images/search`, query),
  byAuthor: author =>
    requests.get(`/notices?author=${encodeURIComponent(author)}&limit=5`),
  favoritedBy: author =>
    requests.get(`/notices?favorited=${encodeURIComponent(author)}&limit=5`),
  get: slug => requests.get(`/notices/${slug}`),
  del: slug => requests.del(`/notices/${slug}`)
};

export default {Images};