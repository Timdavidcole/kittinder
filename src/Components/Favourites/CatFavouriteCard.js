import React from "react";
import agent from "../../agent";

class CatFavouriteCard extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      image: [],
      resolved: false,
      error: true
    };
  }

  componentDidMount() {
    agent.Images.get({ breed_id: this.props.cat.id })
      .then(data =>
        this.setState({
          image: data[0],
          resolved: true,
          error: false
        })
      )
      .catch(error => {
        this.setState({
          error: true
        });
      });
  }

  render() {
    const { cat } = this.props;
    if (this.state.error) {
      return (
        <center>
          <div
            style={{
              backgroudColor: "rgb(255, 182, 205)",
              margin: "20px",
              padding: "5px",
              borderRadius: "20px",
              boxShadow: "0px 0px 15px 0px rgba(158,158,158,1)",
              maxWidth: "400px"
            }}
            id="catFavouriteCard"
          >
            <h2>{cat.name}</h2>
            <div
              id="errorMessage"
              style={{ padding: "5px ", fontSize: "0.8rem" }}
            >
              This cat doesn't like having it's picture shown right now, please
              try again later.
            </div>
            <div style={{ width: "100%" }}>
              <span>{cat.description}</span>
            </div>
          </div>
        </center>
      );
    } else {
      return (
        <center>
          <div
            style={{
              backgroudColor: "rgb(255, 182, 205)",
              margin: "20px",
              padding: "5px",
              borderRadius: "20px",
              boxShadow: "0px 0px 15px 0px rgba(158,158,158,1)",
              maxWidth: "400px"
            }}
            id={`catFavouriteCard`}
          >
            <h2>{cat.name}</h2>
            <img
              id="cat-image"
              src={this.state.image.url}
              alt={"cat"}
              style={{ maxWidth: "400px", maxHeight: "400px" }}
            ></img>
            <div style={{ width: "100%" }}>
              <span>{cat.description}</span>
            </div>
          </div>
        </center>
      );
    }
  }
}

export default CatFavouriteCard;
