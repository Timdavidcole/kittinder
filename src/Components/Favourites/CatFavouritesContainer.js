import React from "react";
import CatFavouriteCard from "./CatFavouriteCard";
import agent from "../../agent";

class CatFavouritesContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      favourites: [],
      error: false
    };
    this.getFavouritesBreedInfo = this.getFavouritesBreedInfo.bind(this);
  }

  getFavouritesBreedInfo(favourites) {
    favourites.forEach(cat => {
      agent.Images.getOne(cat.image_id)
        .then(data =>
          this.setState(prevState => ({
            favourites: [...prevState.favourites, data],
            error: false
          }))
        )
        .catch(() =>
          this.setState({
            error: true
          })
        );
    });
  }

  componentDidMount() {
    agent.Favourites.get(this.props.userId)
      .then(data => this.getFavouritesBreedInfo(data))
      .catch(() =>
        this.setState({
          error: true
        })
      );
  }

  render() {
    if (this.state.error) {
      return (
        <div id={"errorCard"} style={{ padding: "5px ", fontSize: "0.8rem" }}>
          Hmmm, we can't find your favourites. Please try again later.
        </div>
      );
    }
    if (this.state.favourites[0]) {
      return (
        <div className="catCardContainer" style={{ padding: "10px" }}>
          {this.state.favourites.map((cat, index) => {
            return (
              <CatFavouriteCard
                userId={this.props.userId}
                cat={cat.breeds[0]}
                key={index}
                id={index}
              />
            );
          })}
        </div>
      );
    } else return <div></div>;
  }
}

export default CatFavouritesContainer;
