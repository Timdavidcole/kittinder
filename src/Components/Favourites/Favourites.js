import React from "react";
import agent from "../../agent";
import CatFavouritesContainer from "./CatFavouritesContainer";

class Favourites extends React.Component {
  constructor(props) {
    super(props);
    this.state = { error: true };
  }

  componentDidMount() {
    agent.Breeds.all()
      .then(data => {
        this.props.updateCats(data);
        this.setState({ error: false });
      })
      .catch(() =>
        this.setState({
          error: true
        })
      );
  }

  render() {
    if (this.state.error) {
      return (
        <div style={{ padding: "5px ", fontSize: "0.8rem" }}>
          Hmmm, we can't find your favourites.  Please try again later.
        </div>
      );
    } else {
      return (
        <div>
          <CatFavouritesContainer
            userId={this.props.userId}
          />
        </div>
      );
    }
  }
}

export default Favourites;
