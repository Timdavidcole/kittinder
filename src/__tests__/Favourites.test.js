import Favourites from "../Components/Favourites/Favourites";
import React from "react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import { mount, shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

jest.mock("../agent");
import agent from "../agent";

beforeEach(() => {
  agent.Breeds.all.mockImplementation(() => {
    return Promise.resolve([
      { id: "test", name: "testcat", description: "Just a test cat" }
    ]);
  });
  agent.Favourites.get.mockImplementation(() => {
    return Promise.resolve([
      { id: "test", name: "testcat", description: "Just a test cat" }
    ]);
  });
});

test("checks Favourites snapshot", async () => {
  const favourites = await renderer.create(<Favourites />);

  let tree = favourites.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Favourites has correct breeds array from API call", async () => {
  const wrapper = await mount(<Favourites />);
  expect(agent.Breeds.all).toBeCalled();
  await wrapper.instance().hasFinishedAsync;
  wrapper.setProps({ cats: "TEST" });
  expect(wrapper.find("Favourites").props()).toEqual({ cats: "TEST" });
});

test("Favourites shows correct error message on API catch", async () => {
  agent.Breeds.all.mockImplementation(() => {
    return Promise.reject(new Error("fail"));
  });

  const wrapper = await shallow(<Favourites />);
  expect(wrapper.text()).toBe(
    "Hmmm, we can't find your favourites.  Please try again later."
  );
});
