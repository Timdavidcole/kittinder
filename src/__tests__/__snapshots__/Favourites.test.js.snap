// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks Favourites snapshot 1`] = `
<div
  style={
    Object {
      "fontSize": "0.8rem",
      "padding": "5px ",
    }
  }
>
  Hmmm, we can't find your favourites.  Please try again later.
</div>
`;
